File: trwm/reliability.py
from __future__ import annotations

from dataclasses import asdict, dataclass, field, replace
from math import sqrt
from typing import Any, Iterable, Mapping

from .core import Receipt, stable_hash


RELIABILITY_SNAPSHOT_SCHEMA = "trwm.reliability_snapshot.v1"


@dataclass(frozen=True)
class VerifierReliabilityRow:
    subject_id: str
    audited_successes: int = 0
    audited_failures: int = 0
    blocked_unknown: int = 0
    observations: int = 0
    posterior_mean: float = 0.5
    wilson_lower_bound: float = 0.0
    risk_score: float = 1.0


@dataclass(frozen=True)
class VerifierReliabilitySnapshot:
    schema_version: str
    rows: tuple[VerifierReliabilityRow, ...]
    prior_success: float = 1.0
    prior_failure: float = 1.0
    z: float = 1.96
    snapshot_hash: str = ""

    def __post_init__(self) -> None:
        if self.schema_version != RELIABILITY_SNAPSHOT_SCHEMA:
            raise ValueError(f"invalid reliability snapshot schema: {self.schema_version}")
        object.__setattr__(self, "rows", tuple(sorted(self.rows, key=lambda row: row.subject_id)))
        if not self.snapshot_hash:
            object.__setattr__(self, "snapshot_hash", verifier_reliability_snapshot_hash(self))


class VerifierReliabilityMemory:
    def __init__(self, *, prior_success: float = 1.0, prior_failure: float = 1.0, z: float = 1.96) -> None:
        if prior_success <= 0 or prior_failure <= 0:
            raise ValueError("priors must be positive")
        if z <= 0:
            raise ValueError("z must be positive")
        self.prior_success = float(prior_success)
        self.prior_failure = float(prior_failure)
        self.z = float(z)
        self._rows: dict[str, VerifierReliabilityRow] = {}

    def update(self, subject_id: str, *, audited_success: bool | None) -> VerifierReliabilityRow:
        subject = str(subject_id)
        if not subject:
            raise ValueError("subject_id must be non-empty")
        row = self._rows.get(subject, VerifierReliabilityRow(subject_id=subject))
        if audited_success is True:
            row = replace(row, audited_successes=row.audited_successes + 1, observations=row.observations + 1)
        elif audited_success is False:
            row = replace(row, audited_failures=row.audited_failures + 1, observations=row.observations + 1)
        else:
            row = replace(row, blocked_unknown=row.blocked_unknown + 1, observations=row.observations + 1)
        row = self._scored(row)
        self._rows[subject] = row
        return row

    def update_from_receipt(self, receipt: Receipt) -> VerifierReliabilityRow | None:
        residual = receipt.hard_result.residual
        metadata = receipt.hard_result.metadata
        if isinstance(residual, Mapping) and _get(residual, "kind") == "verifier_false_positive":
            subject = _get(residual, "primary_verifier_id", "primaryVerifierId")
            audit_result = str(_get(residual, "audit_result", "auditResult") or "")
            if subject:
                return self.update(str(subject), audited_success=False if audit_result == "reject" else None)
        primary_result = str(_get(metadata, "primary_result", "primaryResult") or "")
        audit_result = str(_get(metadata, "audit_result", "auditResult") or "")
        subject = _get(metadata, "primary_verifier_id", "primaryVerifierId")
        if receipt.hard_result.accepted and primary_result == "accept" and audit_result == "accept" and subject:
            return self.update(str(subject), audited_success=True)
        return None

    def score(self, subject_id: str) -> VerifierReliabilityRow:
        subject = str(subject_id)
        return self._rows.get(subject, self._scored(VerifierReliabilityRow(subject_id=subject)))

    def rank_for_audit(self, subject_ids: Iterable[str]) -> tuple[str, ...]:
        unique_subjects = tuple(dict.fromkeys(str(subject) for subject in subject_ids if str(subject)))
        return tuple(
            row.subject_id
            for row in sorted(
                (self.score(subject) for subject in unique_subjects),
                key=lambda row: (-row.risk_score, -row.audited_failures, row.subject_id),
            )
        )

    def select_for_audit(self, subject_ids: Iterable[str], max_audits: int) -> tuple[str, ...]:
        if not isinstance(max_audits, int) or isinstance(max_audits, bool) or max_audits < 0:
            raise ValueError("max_audits must be a non-negative integer")
        return self.rank_for_audit(subject_ids)[:max_audits]

    def snapshot(self) -> VerifierReliabilitySnapshot:
        return VerifierReliabilitySnapshot(
            schema_version=RELIABILITY_SNAPSHOT_SCHEMA,
            rows=tuple(self._rows.values()),
            prior_success=self.prior_success,
            prior_failure=self.prior_failure,
            z=self.z,
        )

    def _scored(self, row: VerifierReliabilityRow) -> VerifierReliabilityRow:
        successes = row.audited_successes
        failures = row.audited_failures
        n = successes + failures
        posterior_mean = (self.prior_success + successes) / (self.prior_success + self.prior_failure + n)
        lower = wilson_lower_bound(successes, failures, z=self.z)
        risk = 1.0 - lower
        return replace(
            row,
            posterior_mean=_round_float(posterior_mean),
            wilson_lower_bound=_round_float(lower),
            risk_score=_round_float(risk),
        )


def wilson_lower_bound(successes: int, failures: int, *, z: float = 1.96) -> float:
    if successes < 0 or failures < 0:
        raise ValueError("successes and failures must be non-negative")
    n = successes + failures
    if n == 0:
        return 0.0
    phat = successes / n
    z2 = z * z
    denominator = 1.0 + z2 / n
    center = phat + z2 / (2.0 * n)
    margin = z * sqrt((phat * (1.0 - phat) / n) + (z2 / (4.0 * n * n)))
    return max(0.0, (center - margin) / denominator)


def verifier_reliability_snapshot_hash(snapshot: VerifierReliabilitySnapshot) -> str:
    data = asdict(snapshot)
    data.pop("snapshot_hash", None)
    return stable_hash(data)


def validate_verifier_reliability_snapshot(snapshot: VerifierReliabilitySnapshot) -> bool:
    if snapshot.schema_version != RELIABILITY_SNAPSHOT_SCHEMA:
        return False
    if any(row.subject_id == "" for row in snapshot.rows):
        return False
    if len({row.subject_id for row in snapshot.rows}) != len(snapshot.rows):
        return False
    return snapshot.snapshot_hash == verifier_reliability_snapshot_hash(snapshot)


def _get(mapping: Mapping[str, Any], *keys: str) -> Any:
    for key in keys:
        if key in mapping:
            return mapping[key]
    return None


def _round_float(value: float) -> float:
    return round(float(value), 12)

File: src/reliability.ts
import { stableHash } from "./canonical.js";
import type { Receipt } from "./core.js";

export const RELIABILITY_SNAPSHOT_SCHEMA = "trwm.reliability_snapshot.v1";

export interface VerifierReliabilityRow {
  subjectId: string;
  auditedSuccesses: number;
  auditedFailures: number;
  blockedUnknown: number;
  observations: number;
  posteriorMean: number;
  wilsonLowerBound: number;
  riskScore: number;
}

export interface VerifierReliabilitySnapshot {
  schemaVersion: string;
  rows: VerifierReliabilityRow[];
  priorSuccess: number;
  priorFailure: number;
  z: number;
  snapshotHash: string;
}

export class VerifierReliabilityMemory {
  priorSuccess: number;
  priorFailure: number;
  z: number;
  private rows = new Map<string, VerifierReliabilityRow>();

  constructor(options: { priorSuccess?: number; priorFailure?: number; z?: number } = {}) {
    this.priorSuccess = options.priorSuccess ?? 1;
    this.priorFailure = options.priorFailure ?? 1;
    this.z = options.z ?? 1.96;
    if (this.priorSuccess <= 0 || this.priorFailure <= 0) {
      throw new RangeError("priors must be positive");
    }
    if (this.z <= 0) {
      throw new RangeError("z must be positive");
    }
  }

  update(subjectId: string, auditedSuccess: boolean | null): VerifierReliabilityRow {
    const subject = String(subjectId);
    if (!subject) {
      throw new RangeError("subjectId must be non-empty");
    }
    const current = this.rows.get(subject) ?? this.score(subject);
    const pending = { ...current };
    pending.observations += 1;
    if (auditedSuccess === true) {
      pending.auditedSuccesses += 1;
    } else if (auditedSuccess === false) {
      pending.auditedFailures += 1;
    } else {
      pending.blockedUnknown += 1;
    }
    const scored = this.scored(pending);
    this.rows.set(subject, scored);
    return scored;
  }

  updateFromReceipt(receipt: Receipt): VerifierReliabilityRow | null {
    const residual = receipt.hardResult.residual;
    const metadata = receipt.hardResult.metadata;
    if (isRecord(residual) && valueAt(residual, "kind") === "verifier_false_positive") {
      const subject = valueAt(residual, "primaryVerifierId", "primary_verifier_id");
      const auditResult = String(valueAt(residual, "auditResult", "audit_result") ?? "");
      if (subject) {
        return this.update(String(subject), auditResult === "reject" ? false : null);
      }
    }
    const primaryResult = String(valueAt(metadata, "primaryResult", "primary_result") ?? "");
    const auditResult = String(valueAt(metadata, "auditResult", "audit_result") ?? "");
    const subject = valueAt(metadata, "primaryVerifierId", "primary_verifier_id");
    if (receipt.hardResult.result === "accept" && primaryResult === "accept" && auditResult === "accept" && subject) {
      return this.update(String(subject), true);
    }
    return null;
  }

  score(subjectId: string): VerifierReliabilityRow {
    const subject = String(subjectId);
    return this.rows.get(subject) ?? this.scored({
      subjectId: subject,
      auditedSuccesses: 0,
      auditedFailures: 0,
      blockedUnknown: 0,
      observations: 0,
      posteriorMean: 0.5,
      wilsonLowerBound: 0,
      riskScore: 1,
    });
  }

  rankForAudit(subjectIds: string[]): string[] {
    return Array.from(new Set(subjectIds.map(String).filter(Boolean)))
      .map((subject) => this.score(subject))
      .sort((a, b) =>
        b.riskScore - a.riskScore
        || b.auditedFailures - a.auditedFailures
        || compareCodePoint(a.subjectId, b.subjectId)
      )
      .map((row) => row.subjectId);
  }

  selectForAudit(subjectIds: string[], maxAudits: number): string[] {
    if (!Number.isInteger(maxAudits) || maxAudits < 0) {
      throw new RangeError("maxAudits must be a non-negative integer");
    }
    return this.rankForAudit(subjectIds).slice(0, maxAudits);
  }

  async snapshot(): Promise<VerifierReliabilitySnapshot> {
    const pending: VerifierReliabilitySnapshot = {
      schemaVersion: RELIABILITY_SNAPSHOT_SCHEMA,
      rows: Array.from(this.rows.values()).sort((a, b) => compareCodePoint(a.subjectId, b.subjectId)),
      priorSuccess: this.priorSuccess,
      priorFailure: this.priorFailure,
      z: this.z,
      snapshotHash: "",
    };
    return { ...pending, snapshotHash: await verifierReliabilitySnapshotHash(pending) };
  }

  private scored(row: VerifierReliabilityRow): VerifierReliabilityRow {
    const n = row.auditedSuccesses + row.auditedFailures;
    const posteriorMean = (this.priorSuccess + row.auditedSuccesses)
      / (this.priorSuccess + this.priorFailure + n);
    const lower = wilsonLowerBound(row.auditedSuccesses, row.auditedFailures, this.z);
    return {
      ...row,
      posteriorMean: roundFloat(posteriorMean),
      wilsonLowerBound: roundFloat(lower),
      riskScore: roundFloat(1 - lower),
    };
  }
}

export function wilsonLowerBound(successes: number, failures: number, z = 1.96): number {
  if (!Number.isInteger(successes) || !Number.isInteger(failures) || successes < 0 || failures < 0) {
    throw new RangeError("successes and failures must be non-negative integers");
  }
  const n = successes + failures;
  if (n === 0) {
    return 0;
  }
  const phat = successes / n;
  const z2 = z * z;
  const denominator = 1 + z2 / n;
  const center = phat + z2 / (2 * n);
  const margin = z * Math.sqrt((phat * (1 - phat) / n) + (z2 / (4 * n * n)));
  return Math.max(0, (center - margin) / denominator);
}

export async function verifierReliabilitySnapshotHash(snapshot: VerifierReliabilitySnapshot): Promise<string> {
  const { snapshotHash: _snapshotHash, ...withoutHash } = snapshot;
  return stableHash(withoutHash);
}

export async function validateVerifierReliabilitySnapshot(snapshot: VerifierReliabilitySnapshot): Promise<boolean> {
  if (snapshot.schemaVersion !== RELIABILITY_SNAPSHOT_SCHEMA) {
    return false;
  }
  if (snapshot.rows.some((row) => !row.subjectId)) {
    return false;
  }
  if (new Set(snapshot.rows.map((row) => row.subjectId)).size !== snapshot.rows.length) {
    return false;
  }
  return snapshot.snapshotHash === await verifierReliabilitySnapshotHash(snapshot);
}

function valueAt(record: Record<string, unknown>, ...keys: string[]): unknown {
  for (const key of keys) {
    if (Object.prototype.hasOwnProperty.call(record, key)) {
      return record[key];
    }
  }
  return undefined;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

function compareCodePoint(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

function roundFloat(value: number): number {
  return Math.round(value * 1_000_000_000_000) / 1_000_000_000_000;
}
